server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -265,10 +266,28 @@ public CachingClusteredClient(
       }
 
       final Set<ServerToSegment> segments = computeSegmentsToQuery(timeline);
-      @Nullable final byte[] queryCacheKey = computeQueryCacheKey();
+      @Nullable
+      final byte[] queryCacheKey = computeQueryCacheKey();
+
+      if (responseContext.containsKey(CacheConfig.ENABLE_RESULTLEVEL_CACHE)
+          || query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) != null) {
+        @Nullable
+        final String prevEtag = (responseContext.get(QueryResource.EXISTING_RESULT_ID) == null)
+                                ? (String) query.getContext()
+                                                .get(QueryResource.HEADER_IF_NONE_MATCH)
+                                : (String) responseContext.get(QueryResource.EXISTING_RESULT_ID);
+        @Nullable
+        final String currentEtag = computeCurrentEtag(segments, queryCacheKey);
+        if (currentEtag != null && currentEtag.equals(prevEtag)) {
+          return Sequences.empty();
+        }
+      }
```

Comment:
can ResultLevelCachingQueryRunner add  "HEADER_IF_NONE_MATCH" in the responseContext instead of EXISTING_RESULT_ID, so that above computation is not necessary  and check for "ENABLE_RESULTLEVEL_CACHE" would happen in ResultLevelCachingQueryRunner?